File: blind_150_solutions/stack/valid_parentheses/validParentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {

    let stack = [];
    const bracketsMap = new Map([
        [')', '('],
        [']', '['],
        ['}', '{']
    ]);
    
    for (const bracket of s) {
        if (!bracketsMap.has(bracket)) {
            stack.push(bracket);
        } else if (!stack.length || bracketsMap.get(bracket) != stack.pop()) {
            return false;
        }
    }

    return !stack.length ? true : false;
};

File: blind_150_solutions/stack/valid_parentheses/valid_parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        
        stack = []
        closed_brackets = {')':'(', ']':'[', '}': '{'}

        for bracket in s:
            if not bracket in closed_brackets:
                stack.append(bracket)

            elif not stack or closed_brackets[bracket] != stack.pop():
                return False

        return not stack